core/src/main/java/org/infinispan/statetransfer/StateTransferInterceptor.java

```diff
@@ -231,6 +216,24 @@ private Object handleTxCommand(TxInvocationContext ctx, TransactionBoundaryComma
          }
       } else {
          if (currentTopologyId() > command.getTopologyId()) {
+            /*
+             * We need to retry in case of topology-id mismatch.
+             * We rollback the transaction is it is a 2PC. The rollback is safe (even with state transfer prepared transactions)
+            because we ask the coordinator
+            to retry the transaction. The retry will re-acquire the locks and perform the validation needed. If this node
```

Comment:
"we ask the coordinator to retry the transaction" - like I said, this is not always true: if a node just joined, it could receive the `PrepareCommand` asynchronously, after the originator finished the sync broadcast, so it couldn't tell the originator to retry.